src/meta/src/barrier/mod.rs

```diff
@@ -815,61 +831,83 @@ where
+                        synced_ssts.is_empty(),
+                        "no sstables should be produced in the first epoch"
+                    );
+                } else if checkpoint {
+                    self.hummock_manager
+                        .commit_epoch(node.command_ctx.prev_epoch.0, synced_ssts, sst_to_worker)
+                        .await?;
+                } else {
+                    self.hummock_manager.update_current_epoch(prev_epoch)?;
+                    // if we collect a barrier(checkpoint = false),
+                    // we need to ensure that command is Plain and the notifier's checkpoint is
+                    // false
+                    assert_matches!(node.command_ctx.command, Command::Plain(_));
+                }
+
+                node.command_ctx.post_collect().await?;
+
+                // Notify about collected.
+                let mut notifiers = take(&mut node.notifiers);
+                notifiers.iter_mut().for_each(|notifier| {
+                    if !checkpoint {
+                        assert!(!notifier.bound_to_checkpoint_barrier())
                     }
-                    while let Some((command_ctx, mut notifiers, create_mv_progress)) =
-                        uncommitted_states.uncommitted_checkpoint_post.pop_back()
-                    {
-                        checkpoint_control.remove_changes(command_ctx.command.changes());
-                        command_ctx.post_collect().await?;
-
-                        // Notify about collected first.
-                        notifiers.iter_mut().for_each(Notifier::notify_collected);
-
-                        // Then try to finish the barrier for Create MVs.
-                        let actors_to_finish = command_ctx.actors_to_track();
-                        tracker.add(command_ctx.curr_epoch, actors_to_finish, notifiers);
-                        for progress in create_mv_progress {
-                            tracker.update(&progress);
-                        }
+                    notifier.notify_collected();
+                });
+
+                // Save `finished_notifier` for Create MVs.
+                let actors_to_finish = node.command_ctx.actors_to_track();
+                let mut finished_notifiers =
+                    tracker.add(node.command_ctx.curr_epoch, actors_to_finish, notifiers);
+                for progress in resps.iter().flat_map(|r| r.create_mview_progress.clone()) {
+                    if let Some(mut notifier) = tracker.update(&progress) {
+                        finished_notifiers.append(&mut notifier);
                     }
-                } else if prev_epoch != INVALID_EPOCH {
-                    self.hummock_manager.update_current_epoch(prev_epoch)?;
                 }
+
+                // Force checkpoint in next barrier to finish creating mv.
+                if !finished_notifiers.is_empty() && !checkpoint {
+                    checkpoint_control.force_checkpoint_in_next_barrier();
```

Comment:
So creating an empty materialized view will last for 2 barrier intervals in most cases? 🤣 This may make the e2e tests much slower in CI. How about decreasing the barrier interval?

Reply:
e2e compared to other PRs. 🥵
<img width="1166" alt="image" src="https://user-images.githubusercontent.com/25862682/191672647-4a3d5c71-3876-46aa-bb92-2d57285dfe7d.png">
<img width="1161" alt="image" src="https://user-images.githubusercontent.com/25862682/191672668-d3c9f6e4-165f-4404-bfb6-46ba4c7c2637.png">

Reply:
But barrier intervals can be further decreased?

Reply:
I have found other pr's that also take more than 9 minutes to run. So maybe this is not the main reason

Reply:
How is this resolved now?